docs/specification/2025-03-26/basic/authorization.mdx

```diff
@@ -8,303 +8,466 @@ title: Authorization
+  { "issuer":"https://authserver2.example", "client_id":"urn:client_ids:546"}
+]
+```
 
-The complete Authorization flow proceeds as follows:
+Alternatively, an MCP client could allow the user to register the MCP client and provide those details
+to the MCP client through a UI.
+
+### 2.8 Authorization flow steps
+
+The complete authorization flow proceeds as follows:  
 
 ```mermaid
 sequenceDiagram
-    participant B as User-Agent (Browser)
-    participant C as Client
-    participant M as MCP Server
-
-    C->>M: GET /.well-known/oauth-authorization-server
-    alt Server Supports Discovery
-        M->>C: Authorization Server Metadata
-    else No Discovery
-        M->>C: 404 (Use default endpoints)
+    participant UA as User-Agent (Browser or App/Broker)
+    participant MC as MCP Client
+    participant MS as MCP Server
+    participant AS as Authorization Server
+
+    MC->>MS: Request data
+    MS-->>MC: 401 Unauthorized with WWW-Authenticate
+
+    MC->>MS: Fetch PRM document
+    MS-->>MC: PRM document response
+    Note over MC: Validate PRM document<br/>metadata
+    Note over MC: Build AS metadata URL
+    MC->>AS: Fetch AS metadata
+    AS-->>MC: AS metadata response
+
+    alt External authorization server
+        MC->>AS: GET /.well-known/oauth-authorization-server
+        AS-->>MC: 200 OK + Metadata Document
+        Note over MC: Use endpoints from metadata
+    else Co-located authorization server
+        MC->>MS: GET /.well-known/oauth-authorization-server
+        MS-->>MC: 200 OK + Metadata Document
+        Note over MC: Use endpoints from metadata
     end
 
-    alt Dynamic Client Registration
-        C->>M: POST /register
-        M->>C: Client Credentials
+    alt Client is pre-registered with the authorization server
+    else Authorization server supports dynamic client registration
+        MC->>AS: Request client registration
+        AS-->>MC: Confirm client registration
+    else Authorization server does not support dynamic client registration
+        MC->>MS: Request client registration
+        MS->>AS: Register client
+        AS-->>MS: Return client registration
+        MS-->>MC: Return client registration
     end
 
-    Note over C: Generate PKCE Parameters
-    C->>B: Open browser with authorization URL + code_challenge
-    B->>M: Authorization Request
-    Note over M: User /authorizes
-    M->>B: Redirect to callback with authorization code
-    B->>C: Authorization code callback
-    C->>M: Token Request + code_verifier
-    M->>C: Access Token (+ Refresh Token)
-    C->>M: API Requests with Access Token
+    Note over MC: Generate PKCE artifacts
+    MC-->>UA: Initiate authorization call
+    Note over UA: User authenticates and authorizes<br/>client for scope and resource
+    UA->>MC: Return auth code to client
+    MC->>AS: Exchange auth code for tokens
+    AS-->>MC: Return tokens
+    Note over MC: Store tokens
+    Note over MC, MS: Begin standard MCP message exchange
```

Comment:
Should we mention explicitly that the messages include the token? This is already noted in 2.4.2, and it's reasonable to assume it's understood here, but perhaps we should duplicate that here.

Related: "begin" might imply something about the initial SSE connection and not subsequent requests. I mention it because it seems some clients only use configured headers/QSPs on the initial connection, but drop them for subsequent messages. That's probably a bug on their end, but it still makes me think additional prescriptivism around using the token for all messages might be beneficial.

Something similar might be good to have in 2.9 as well.